src/db/pool.cr

```diff
@@ -47,6 +60,56 @@ module DB
       @initial_pool_size.times { build_resource }
     end
 
+    def reap_connection!(connection : T, delay : Float64)
+      return unless connection.responds_to?(:check)
+
+      sleep delay
+
+      sync do
+        connection_idle = @idle.includes? connection
+        return unless connection_idle # let the next reaping attempt get it
+
+        delete connection
+      end
+
+      connection.check
+
+      sync do
+        @total << connection
+        @idle << connection
+      end
+    rescue exc : DB::ConnectionLost
+      connection.close
+
+      sync do
+        add_resource! if can_increase_pool?
+      end
+    end
+
+    def add_resource!
+      @inflight += 1
+      r = unsync { build_resource }
+    ensure # prevent inflight from incrementing if build_resource fails
+      @inflight -= 1
+      r
+    end
+
+    def start_reaper!
+      spawn do
+        sleep @reaping_delay
+
+        loop do
```

Comment:
This fiber should be stopped when the database gets closed.

Reply:
It feels like `sleep @reaping_frequency` would throw that loop off, so I may still be missing something here -- this was inspired by https://lbarasti.com/post/select_statement/ but I definitely don't fully grok it yet.

Reply:
You can use `when timeout(@reaping_frequency)` instead of `else`. This means: What happens first, the timeout completes or the channel receives, the respective branch executes.

Reply:
Done, thanks!